tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp

```diff
@@ -2794,6 +2794,56 @@ unsigned AlignBufferOffsetInLegacy(unsigned offset, unsigned size,
   return offset;
 }
 
+// Translate HitObject constructor to dx.op.hitObject_MakeNop
+void TranslateHitObjectConstructor(HLModule &HLM) {
```

Comment:
This concerns me. CGHLSLMS is a giant hack that is nothing but a problem.

Is there a reason we can't have the AST represent a sane constructor that we can codegen directly?

Reply:
> With the NodeRecord types we demonstrated a new solution, where we add an unspellable attribute to the type that identifies it as a specific object. We should do something like that here as well.

bf3400a99 follows your suggestion and adds the `HLSLHitObjectAttr` attribute to model the HitObject type analogously to how it's done with `GetNodeIOType`

> This concerns me. CGHLSLMS is a giant hack that is nothing but a problem.
> Is there a reason we can't have the AST represent a sane constructor that we can codegen directly?

Luckily, `EmitHLSLScalarObjectDefaultConstructor` directly emits a call to `HitObject_MakeNop`. This change in CGHLSLMS was never needed and i removed it in 29c8705e5.

Reply:
This is obsolete, 358942da6 removes the special handling for the HitObject type in CGHLSLMS and elsewhere.